app/interactors/featured_attachments/update_order_interactor.rb

```diff
@@ -0,0 +1,45 @@
+class FeaturedAttachments::UpdateOrderInteractor < ApplicationInteractor
+  delegate :params,
+           :user,
+           :edition,
+           :attachments,
+           :updater,
+           to: :context
+
+  def call
+    Edition.transaction do
+      find_and_lock_edition
+      reorder_attachments
+      update_edition
+      create_timeline_entry
+    end
+  end
+
+  def find_and_lock_edition
+    context.edition = Edition.lock.find_current(document: params[:document])
+    assert_edition_state(edition, &:editable?)
+
+    assert_edition_feature(edition, assertion: "supports featured attachments") do
+      edition.document_type.attachments.featured?
+    end
+  end
+
+  def reorder_attachments
+    new_ordering = params[:ordering].keys.sort_by do |featured_attachment_id|
+      params[:ordering][featured_attachment_id].to_i
+    end
```

Comment:
This seems to assume that a users query included all attachments in it (and doesn't include any junk) which could be a source of subtle and rare bugs.

I'd suggest sorting a duplicate of the existing featured_attachment_ordering and looking up the param of each one so that way it'll contain all the same items only re-ordered and won't allow anything unexpected in there.

Reply:
This is an interesting one. I actually used your suggested approach originally, and then found this to be slightly less convoluted. Some reasons this could be acceptable:

   - Ordering is 'optional', in that we cope without it
   - If there is dodgy data, we'll store it and be able to debug
   - The validation protects us from obviously dodgy stuff

Would you be more concerned if we continued as-is?

Reply:
Yeah, I think this way seems the less ideal approach since it would miss any new items added during submission and any dodgy items would raise a 500 exception rather than be a client error.

Reply:
Ah, fair enough - I thought it would give a 4xx response. Updated to ensure we only set a valid ordering.

Reply:
Thanks - one thing to note is that to_i will mean that anything not submitted will end up at the top of the list whereas Float::INFINITY would push those to the bottom. I feel the latter is slightly more intuitive but not too big a deal as long as we've made an informed choice.

Also I don't think there's anything that ensures this is a hash so this could still crash off unexpected input. Perhaps we can use dig on params to avoid that?

Reply:
Indeed, the `to_i` seemed like the simplest thing to ensure we "cope" with that scenario (rather than explaining the need for a specific position). I've updated the commits to explain this more; we could always define the ordering differently if it becomes an issue.

I was about to follow the `dig` suggestion, but then realised `[].dig` would cause a crash. In this case, I think we may be putting too much thought into something that should only affect the user submitting the form, and is likely due to them doing some form of hackery.

Is there a particular scenario you think we should protect against here?

Reply:
I think it'll only be an issue if multiple users work on the same thing and are surprised by the ordering, which is hopefully a very low risk.

Rather than particular scenarios we should just always be cautious that user input is handled in the appropriate way - so any case where a user input can cause a 500 error is a smell to me. 

I did some tinkering and realise dig isn't the right approach apologies, it does leave vulnerability for 500s still. Hmm the Rails preferred approach would be using the `params.permit(ordering: {})` which raises a bad request if it's not a hash - however this creates that annoying second warning for unpermitted params - so this would probably better in a context of `params.require(:attachments).permit(ordering: {})` which would be a clean way with bad requests handled.

But yeah without making further modifications this could be safe with `params[:ordering].try(:[], attachment.featured_attachment_id).to_i`.

Reply:
Hmm...that doesn't work if we do this:

```
irb(main):017:0> ActionController::Parameters.new(ordering: [1])[:ordering].try(:[], :foo)      
Traceback (most recent call last):
        2: from (irb):17
        1: from (irb):17:in `rescue in irb_binding'
TypeError (no implicit conversion of Symbol into Integer)
```

I agree it makes sense to have definite knowledge about what our app would do in the case of an invalid input. I've added an extra commit in order to do a `require...permit`.

Could you take another look?

Reply:
Argh good catch! thanks for that. Looks good 👍 I think the `to_h` is redundant in `ordering_params` but I know we do similar elsewhere to break out of Parameters so no biggie.